language/bytecode-verifier/src/type_memory_safety.rs

```diff
@@ -126,52 +123,55 @@ impl<'a> TypeAndMemorySafetyAnalysis<'a> {
         state: &mut AbstractState,
         offset: usize,
         mut_: bool,
-        field_definition_index: FieldDefinitionIndex,
+        field_handle_index: FieldHandleIndex,
+        type_args: &Signature,
     ) {
+        // load operand and check mutability constraints
         let operand = self.stack.pop().unwrap();
-        let struct_handle_index =
-            match SignatureToken::get_struct_handle_from_reference(&operand.signature) {
-                Some(struct_handle_index) => struct_handle_index,
-                None => {
-                    errors.push(err_at_offset(
-                        StatusCode::BORROWFIELD_TYPE_MISMATCH_ERROR,
-                        offset,
-                    ));
-                    return;
-                }
-            };
-        if !self
-            .module()
-            .is_field_in_struct(field_definition_index, struct_handle_index)
-        {
+        if mut_ && !operand.signature.is_mutable_reference() {
             errors.push(err_at_offset(
-                StatusCode::BORROWFIELD_BAD_FIELD_ERROR,
+                StatusCode::BORROWFIELD_TYPE_MISMATCH_ERROR,
                 offset,
             ));
             return;
         }
-        if mut_ && !operand.signature.is_mutable_reference() {
+
+        // check the reference on the stack is the expected type.
+        // Load the type that owns the field according to the instruction.
+        // For generic fields access, this step materializes that type
+        let field_handle = self.module().field_handle_at(field_handle_index);
+        let struct_def = self.module().struct_def_at(field_handle.owner);
+        let expected_type = materialize_type(struct_def.struct_handle, type_args);
+        if !operand.reference_to(&expected_type) {
             errors.push(err_at_offset(
                 StatusCode::BORROWFIELD_TYPE_MISMATCH_ERROR,
                 offset,
             ));
             return;
         }
 
-        if let Some(id) = state.borrow_field(&operand, mut_, field_definition_index) {
-            let field_signature = self
-                .module()
-                .get_field_signature(field_definition_index)
-                .0
-                .clone();
-            let field_token = Box::new(
-                field_signature
-                    .substitute(operand.signature.get_type_actuals_from_reference().unwrap()),
-            );
+        if let Some(id) = state.borrow_field(&operand, mut_, field_handle_index) {
+            // load the field type
+            let field_def = match &struct_def.field_information {
+                StructFieldInformation::Native => {
+                    errors.push(err_at_offset(
+                        StatusCode::BORROWFIELD_BAD_FIELD_ERROR,
+                        offset,
+                    ));
+                    return;
+                }
```

Comment:
not sure... and I mean in the past we had the field definition table and the checks were done differently.
I don't think this is a good place for this check and it may even be (though I cannot find it right now) that the check is elsewhere too and this could be an `unreachable`. But while writing the code I did not want to "crash" and wrote it that way.
This needs some review along the other spot and I'd like to do that at a later time as I think checks and their order in the verifier may be reviewed...